enhancements/insights/insights-runtime-extractor.md

```diff
@@ -0,0 +1,647 @@
+
+The `exporter` component (that runs with no privileges) handles reading the input from the fingerprints execution from the shared volume.
+
+#### Privacy Risk & Mitigation
+
+As runtime information is extracted from the user containers, it is technically possible to get access to sensitive information that should not be exposed to Red Hat.
+The fingerprints that are executing in the container namespaces must be able to sanitize the extracted information and discard them if they deviate from the expected output.
+
+
+To make sure customer confidentiality is preserved, any new collected data is hashed before they are sent to Red Hat backends. Mechanisms are already in place so that access to the unobfuscated data in Red Hat is granted on a needs-to-know basis.
+
+#### Performance Risk & Mitigation
+
+##### Limited Restrictions for Fingerprints
+
+Since a fingerprint executable runs under a downgraded privilege model, with a limited process life-span, significantly more freedom is afforded to it, but some care is still necessary. Notably, a fingerprint executable can have a larger dependency set beyond the core language SDK to meet its needs. Additionally, fingerprint executables can leverage languages with a greater emphasis on productivity (e.g. Golang), provided the implementations still utilize input and output hardening techniques, and do not load potentially rogue code in the container. Each fingerprint implementation may even have a different language runtime than its peers. The following rules should be followed:
+
+1. __Inspecting code must be statically compiled and not rely on dynamic library loading.__ This is essentially already required to achieve stable execution on arbitrary container layers, including those with differing libc implementations, or even scratch containers with no additional libraries
+2. __All analysis I/O operations should be limit-bound.__ A container either through error or design may have files of arbitrary size, or files in a corrupt state. Any potentially unbounded operation (e.g. reading to the end of a file or following a reference data structure to its arbitrary end) is prohibited.
+3. __Memory and CPU usage should not be input derived.__ In addition to achieving system stability and performance goals, ensuring the overhead of a inspecting process does not increase significantly relative to input data is critical in preventing potential Denial-of-Service attacks relying on a specially crafted container filesystem.  
+
+##### No Impact on Workloads
+
+Inspecting containers must not disrupt the containers that are inspected and have an impact on their quality of service.
+The insights-runtime-extractor is running with its own memory and CPU requirements and does not consume resources from the inspected container.
+
+###### Risk of misbehaving code execution
+
+A few fingerprints relies on executing process in the user namespace (for example, executing `node --version` to determine the version of Node.js).
```

Comment:
Starting a lot of processes at the same time for all pods could lead to node instability if node is already near limits. Is there any plan to spread out the scan over some period of time?

Reply:
I updated the `No Impact on Workloads` section above to clarify the behaviour:

> If the worker node is near limits, a burst of activity from the insights runtime extractor could lead to node instability. To prevent this, the insights runtime extractor will extract runtime information from processes one at a time and will not parallelize these tasks.
 Increasing the gathering time is a trade-off to prevent node instability.

Does that answer your question?